src/codegen/llvm/llvm_utils.cpp

```diff
@@ -61,6 +63,85 @@ static void run_optimisation_passes(llvm::Module& module,
 /*                             Optimisation utils                                       */
 /****************************************************************************************/
 
+void initialise_nvptx_passes() {
+    // Register targets.
+    LLVMInitializeNVPTXTarget();
+    LLVMInitializeNVPTXTargetMC();
+    LLVMInitializeNVPTXTargetInfo();
+    LLVMInitializeNVPTXAsmPrinter();
+
+    // Initialize passes.
+    initialise_optimisation_passes();
+}
+
+void optimise_module_for_nvptx(codegen::Platform& platform,
+                               llvm::Module& module,
+                               int opt_level,
+                               std::string& target_asm) {
+    // CUDA target machine we generating code for.
+    std::unique_ptr<llvm::TargetMachine> tm;
+    std::string platform_name = platform.get_name();
+
+    // Target and layout information.
+    static const std::map<std::string, std::string> triple_str = {
+            {"nvptx", "nvptx-nvidia-cuda"},
+            {"nvptx64", "nvptx64-nvidia-cuda"}};
+    static const std::map<std::string, std::string> data_layout_str = {
+            {"nvptx", "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32"
+                      "-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v32:32:32"
+                      "-v64:64:64-v128:128:128-n16:32:64"},
+            {"nvptx64", "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32"
+                        "-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v32:32:32"
+                        "-v64:64:64-v128:128:128-n16:32:64"}};
+
+    // Set data layout and target triple information for the module.
+    auto triple = triple_str.at(platform_name);
+    module.setDataLayout(data_layout_str.at(platform_name);
```

Comment:
I missed closing a parenthesis here :innocent: 
```suggestion
    module.setDataLayout(data_layout_str.at(platform_name));
```